gtsam/hybrid/HybridGaussianFactorGraph.cpp

```diff
@@ -130,49 +128,54 @@ GaussianFactorGraphTree HybridGaussianFactorGraph::assembleGraphTree() const {
 }
 
 /* ************************************************************************ */
-static std::pair<HybridConditional::shared_ptr, HybridFactor::shared_ptr>
+static std::pair<HybridConditional::shared_ptr, boost::shared_ptr<Factor>>
 continuousElimination(const HybridGaussianFactorGraph &factors,
                       const Ordering &frontalKeys) {
   GaussianFactorGraph gfg;
-  for (auto &fp : factors) {
-    if (auto ptr = boost::dynamic_pointer_cast<HybridGaussianFactor>(fp)) {
-      gfg.push_back(ptr->inner());
-    } else if (auto ptr = boost::static_pointer_cast<HybridConditional>(fp)) {
-      gfg.push_back(
-          boost::static_pointer_cast<GaussianConditional>(ptr->inner()));
+  for (auto &f : factors) {
+    if (auto gf = dynamic_pointer_cast<GaussianFactor>(f)) {
+      gfg.push_back(gf);
+    } else if (auto orphan = dynamic_pointer_cast<OrphanWrapper>(f)) {
+      // Ignore orphaned clique.
+      // TODO(dellaert): is this correct? If so explain here.
+    } else if (auto hc = dynamic_pointer_cast<HybridConditional>(f)) {
+      auto gc = hc->asGaussian();
+      if (!gc) throwRuntimeError("continuousElimination", gc);
+      gfg.push_back(gc);
     } else {
-      // It is an orphan wrapped conditional
+      throwRuntimeError("continuousElimination", f);
     }
   }
 
   auto result = EliminatePreferCholesky(gfg, frontalKeys);
-  return {boost::make_shared<HybridConditional>(result.first),
-          boost::make_shared<HybridGaussianFactor>(result.second)};
+  return {boost::make_shared<HybridConditional>(result.first), result.second};
 }
 
 /* ************************************************************************ */
-static std::pair<HybridConditional::shared_ptr, HybridFactor::shared_ptr>
+static std::pair<HybridConditional::shared_ptr, boost::shared_ptr<Factor>>
 discreteElimination(const HybridGaussianFactorGraph &factors,
                     const Ordering &frontalKeys) {
   DiscreteFactorGraph dfg;
 
-  for (auto &factor : factors) {
-    if (auto p = boost::dynamic_pointer_cast<HybridDiscreteFactor>(factor)) {
-      dfg.push_back(p->inner());
-    } else if (auto p = boost::static_pointer_cast<HybridConditional>(factor)) {
-      auto discrete_conditional =
-          boost::static_pointer_cast<DiscreteConditional>(p->inner());
-      dfg.push_back(discrete_conditional);
+  for (auto &f : factors) {
+    if (auto dtf = dynamic_pointer_cast<DecisionTreeFactor>(f)) {
+      dfg.push_back(dtf);
+    } else if (auto orphan = dynamic_pointer_cast<OrphanWrapper>(f)) {
+      // Ignore orphaned clique.
+      // TODO(dellaert): is this correct? If so explain here.
```

Comment:
OrphanWrapper is one of the things that seems a bit "non-canonical" to me in the elimination algorithm.

Reply:
Define non-canonical please?

Reply:
Seems like a hack